File: volley-api/internal/models/game.go
package models

import "time"

// GameCategory represents the sport category
type GameCategory string

const (
	GameCategorySoccer          GameCategory = "soccer"
	GameCategoryBasketball      GameCategory = "basketball"
	GameCategoryPickleball      GameCategory = "pickleball"
	GameCategoryFlagFootball    GameCategory = "flag_football"
	GameCategoryVolleyball      GameCategory = "volleyball"
	GameCategoryUltimateFrisbee GameCategory = "ultimate_frisbee"
	GameCategoryTennis          GameCategory = "tennis"
	GameCategoryOther           GameCategory = "other"
)

// GameStatus represents the current status of a game
type GameStatus string

const (
	GameStatusOpen       GameStatus = "open"        // Accepting sign-ups
	GameStatusFull       GameStatus = "full"        // Roster full, waitlist only
	GameStatusClosed     GameStatus = "closed"      // Sign-up deadline passed
	GameStatusInProgress GameStatus = "in_progress" // Game is happening
	GameStatusCompleted  GameStatus = "completed"   // Game finished
	GameStatusCancelled  GameStatus = "cancelled"   // Game cancelled
)

// SkillLevel represents the skill level requirement for a game
type SkillLevel string

const (
	SkillLevelBeginner     SkillLevel = "beginner"
	SkillLevelIntermediate SkillLevel = "intermediate"
	SkillLevelAdvanced     SkillLevel = "advanced"
	SkillLevelAll          SkillLevel = "all"
)

// PricingType represents how the game is priced
type PricingType string

const (
	PricingTypeFree      PricingType = "free"       // No cost
	PricingTypeTotal     PricingType = "total"      // Total amount split among participants
	PricingTypePerPerson PricingType = "per_person" // Fixed price per person
)

// ParticipantStatus represents the status of a participant
type ParticipantStatus string

const (
	ParticipantStatusConfirmed ParticipantStatus = "confirmed" // Confirmed participant
	ParticipantStatusWaitlist  ParticipantStatus = "waitlist"  // On waitlist
	ParticipantStatusDropped   ParticipantStatus = "dropped"   // Dropped from game
	ParticipantStatusDeclined  ParticipantStatus = "declined"  // Declined invitation
	ParticipantStatusRemoved   ParticipantStatus = "removed"   // Removed from game
)

// Location represents the location details of a game
type Location struct {
	Name      string   `json:"name"`                // Venue or field name
	Address   *string  `json:"address,omitempty"`   // Street address
	Latitude  *float64 `json:"latitude,omitempty"`  // Latitude coordinate
	Longitude *float64 `json:"longitude,omitempty"` // Longitude coordinate
	Notes     *string  `json:"notes,omitempty"`     // Additional location details
}

// Pricing represents the pricing details of a game
type Pricing struct {
	Type        PricingType `json:"type"`        // Pricing type
	AmountCents int         `json:"amountCents"` // Amount in cents (0 for free)
	Currency    string      `json:"currency"`    // Currency code (default: USD)
}

// User represents a basic user structure
type User struct {
	ID        string    `json:"id"`                  // User UUID
	Email     string    `json:"email"`               // User email
	FirstName string    `json:"firstName"`           // User first name
	LastName  string    `json:"lastName"`            // User last name
	CreatedAt time.Time `json:"createdAt,omitempty"` // Account creation timestamp
}

// Team represents a team in a game
type Team struct {
	ID        string    `json:"id"`              // Team UUID
	GameID    string    `json:"gameId"`          // Game UUID this team belongs to
	Name      string    `json:"name"`            // Team name
	Color     *string   `json:"color,omitempty"` // Hex color code
	CreatedAt time.Time `json:"createdAt"`       // Team creation timestamp
}

// Participant represents a user's participation in a game
type Participant struct {
	User                                 // Embedded user (id, email, name, createdAt)
	TeamID             *string           `json:"teamId,omitempty"`             // Team UUID (if assigned)
	Status             ParticipantStatus `json:"status"`                       // Participant status
	WaitlistPosition   *int              `json:"waitlistPosition,omitempty"`   // Position in waitlist
	Paid               bool              `json:"paid"`                         // Payment status
	PaymentAmountCents *int              `json:"paymentAmountCents,omitempty"` // Amount paid in cents
	Notes              *string           `json:"notes,omitempty"`              // Additional notes
	JoinedAt           time.Time         `json:"joinedAt"`                     // When they joined
	UpdatedAt          time.Time         `json:"updatedAt"`                    // Last update timestamp
}

// GameSummary represents essential game details for list views
type GameSummary struct {
	ID                      string             `json:"id"`                                // Game UUID
	Category                GameCategory       `json:"category"`                          // Sport category
	Title                   *string            `json:"title,omitempty"`                   // Custom title
	Description             *string            `json:"description,omitempty"`             // Game description
	Location                Location           `json:"location"`                          // Location details
	StartTime               time.Time          `json:"startTime"`                         // Game start time
	DurationMinutes         int                `json:"durationMinutes"`                   // Duration in minutes
	MaxParticipants         int                `json:"maxParticipants"`                   // Maximum number of players
	SignupCount             int                `json:"signupCount"`                       // Number of participants signed up
	Pricing                 Pricing            `json:"pricing"`                           // Pricing details
	SignupDeadline          time.Time          `json:"signupDeadline"`                    // Sign-up deadline
	SkillLevel              SkillLevel         `json:"skillLevel"`                        // Required skill level
	Status                  GameStatus         `json:"status"`                            // Current game status
	UserParticipationStatus *ParticipantStatus `json:"userParticipationStatus,omitempty"` // Current user's participation status (if authenticated)
}

// Game represents a pickup sports game with full details
type Game struct {
	ID                    string        `json:"id"`                              // Game UUID
	Owner                 *User         `json:"owner,omitempty"`                 // Owner user details
	Category              GameCategory  `json:"category"`                        // Sport category
	Title                 *string       `json:"title,omitempty"`                 // Custom title
	Description           *string       `json:"description,omitempty"`           // Game description
	Location              Location      `json:"location"`                        // Location details
	StartTime             time.Time     `json:"startTime"`                       // Game start time
	DurationMinutes       int           `json:"durationMinutes"`                 // Duration in minutes
	MaxParticipants       int           `json:"maxParticipants"`                 // Maximum number of players
	ConfirmedParticipants []Participant `json:"confirmedParticipants,omitempty"` // Confirmed participants (up to max)
	Waitlist              []Participant `json:"waitlist,omitempty"`              // Waitlisted participants (beyond max)
	Pricing               Pricing       `json:"pricing"`                         // Pricing details
	SignupDeadline        time.Time     `json:"signupDeadline"`                  // Sign-up deadline
	DropDeadline          *time.Time    `json:"dropDeadline,omitempty"`          // Drop deadline (optional)
	SkillLevel            SkillLevel    `json:"skillLevel"`                      // Required skill level
	Notes                 *string       `json:"notes,omitempty"`                 // Additional notes
	Status                GameStatus    `json:"status"`                          // Current game status
	CancelledAt           *time.Time    `json:"cancelledAt,omitempty"`           // When the game was cancelled
	CreatedAt             time.Time     `json:"createdAt"`                       // Creation timestamp
	UpdatedAt             time.Time     `json:"updatedAt"`                       // Last update timestamp
}

// CreateGameRequest represents a request to create a new game
type CreateGameRequest struct {
	Category        GameCategory `json:"category" binding:"required"`               // Sport category
	Title           *string      `json:"title,omitempty"`                           // Custom title
	Description     *string      `json:"description,omitempty"`                     // Game description
	Location        Location     `json:"location" binding:"required"`               // Location details
	StartTime       time.Time    `json:"startTime" binding:"required"`              // Game start time
	DurationMinutes int          `json:"durationMinutes" binding:"required,min=15"` // Duration in minutes
	MaxParticipants int          `json:"maxParticipants" binding:"required,min=2"`  // Maximum number of players
	Pricing         Pricing      `json:"pricing" binding:"required"`                // Pricing details
	SignupDeadline  *time.Time   `json:"signupDeadline,omitempty"`                  // Sign-up deadline (defaults to start_time)
	DropDeadline    *time.Time   `json:"dropDeadline,omitempty"`                    // Drop deadline (optional)
	SkillLevel      *SkillLevel  `json:"skillLevel,omitempty"`                      // Required skill level (defaults to "all")
	Notes           *string      `json:"notes,omitempty"`                           // Additional notes
}

// ListGamesResponse represents the response for listing games
type ListGamesResponse struct {
	Games []GameSummary `json:"games"` // List of game summaries
}

// UpdateGameRequest represents a request to update an existing game
type UpdateGameRequest struct {
	Title           *string     `json:"title,omitempty"`                                      // Custom title
	Description     *string     `json:"description,omitempty"`                                // Game description
	Location        *Location   `json:"location,omitempty"`                                   // Location details
	StartTime       *time.Time  `json:"startTime,omitempty"`                                  // Game start time
	DurationMinutes *int        `json:"durationMinutes,omitempty" binding:"omitempty,min=15"` // Duration in minutes
	MaxParticipants *int        `json:"maxParticipants,omitempty" binding:"omitempty,min=2"`  // Maximum number of players
	Pricing         *Pricing    `json:"pricing,omitempty"`                                    // Pricing details
	SignupDeadline  *time.Time  `json:"signupDeadline,omitempty"`                             // Sign-up deadline
	SkillLevel      *SkillLevel `json:"skillLevel,omitempty"`                                 // Required skill level
	Notes           *string     `json:"notes,omitempty"`                                      // Additional notes
	Status          *GameStatus `json:"status,omitempty"`                                     // Game status
}

File: volley_client/lib/models/game.dart
import 'package:intl/intl.dart';
import 'user.dart';

enum GameCategory {
  soccer,
  basketball,
  pickleball,
  flagFootball,
  volleyball,
  ultimateFrisbee,
  tennis,
  other;

  String get displayName {
    switch (this) {
      case GameCategory.soccer:
        return 'Soccer';
      case GameCategory.basketball:
        return 'Basketball';
      case GameCategory.pickleball:
        return 'Pickleball';
      case GameCategory.flagFootball:
        return 'Football';
      case GameCategory.volleyball:
        return 'Volleyball';
      case GameCategory.ultimateFrisbee:
        return 'Ultimate Frisbee';
      case GameCategory.tennis:
        return 'Tennis';
      case GameCategory.other:
        return 'Other';
    }
  }

  static GameCategory fromString(String value) {
    switch (value) {
      case 'soccer':
        return GameCategory.soccer;
      case 'basketball':
        return GameCategory.basketball;
      case 'pickleball':
        return GameCategory.pickleball;
      case 'flag_football':
        return GameCategory.flagFootball;
      case 'volleyball':
        return GameCategory.volleyball;
      case 'ultimate_frisbee':
        return GameCategory.ultimateFrisbee;
      case 'tennis':
        return GameCategory.tennis;
      default:
        return GameCategory.other;
    }
  }

  String toApiString() {
    switch (this) {
      case GameCategory.soccer:
        return 'soccer';
      case GameCategory.basketball:
        return 'basketball';
      case GameCategory.pickleball:
        return 'pickleball';
      case GameCategory.flagFootball:
        return 'flag_football';
      case GameCategory.volleyball:
        return 'volleyball';
      case GameCategory.ultimateFrisbee:
        return 'ultimate_frisbee';
      case GameCategory.tennis:
        return 'tennis';
      case GameCategory.other:
        return 'other';
    }
  }
}

enum GameStatus {
  open,
  full,
  closed,
  inProgress,
  completed,
  cancelled;

  static GameStatus fromString(String value) {
    switch (value) {
      case 'open':
        return GameStatus.open;
      case 'full':
        return GameStatus.full;
      case 'closed':
        return GameStatus.closed;
      case 'in_progress':
        return GameStatus.inProgress;
      case 'completed':
        return GameStatus.completed;
      case 'cancelled':
        return GameStatus.cancelled;
      default:
        return GameStatus.open;
    }
  }
}

enum PricingType {
  free,
  total,
  perPerson;

  static PricingType fromString(String value) {
    switch (value) {
      case 'free':
        return PricingType.free;
      case 'total':
        return PricingType.total;
      case 'per_person':
        return PricingType.perPerson;
      default:
        return PricingType.free;
    }
  }
}

enum SkillLevel {
  beginner,
  intermediate,
  advanced,
  all;

  String get displayName {
    switch (this) {
      case SkillLevel.beginner:
        return 'Beginner';
      case SkillLevel.intermediate:
        return 'Intermediate';
      case SkillLevel.advanced:
        return 'Advanced';
      case SkillLevel.all:
        return 'All Levels';
    }
  }

  static SkillLevel fromString(String value) {
    switch (value) {
      case 'beginner':
        return SkillLevel.beginner;
      case 'intermediate':
        return SkillLevel.intermediate;
      case 'advanced':
        return SkillLevel.advanced;
      default:
        return SkillLevel.all;
    }
  }
}

class Location {
  final String name;
  final String? address;
  final double? latitude;
  final double? longitude;
  final String? notes;

  Location({
    required this.name,
    this.address,
    this.latitude,
    this.longitude,
    this.notes,
  });

  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      name: json['name'] as String,
      address: json['address'] as String?,
      latitude: json['latitude'] != null
          ? (json['latitude'] as num).toDouble()
          : null,
      longitude: json['longitude'] != null
          ? (json['longitude'] as num).toDouble()
          : null,
      notes: json['notes'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      if (address != null) 'address': address,
      if (latitude != null) 'latitude': latitude,
      if (longitude != null) 'longitude': longitude,
      if (notes != null) 'notes': notes,
    };
  }
}

class Pricing {
  final PricingType type;
  final int? amountCents;
  final String? currency;

  Pricing({
    required this.type,
    this.amountCents,
    this.currency,
  });

  factory Pricing.fromJson(Map<String, dynamic> json) {
    return Pricing(
      type: PricingType.fromString(json['type'] as String),
      amountCents: json['amountCents'] != null
          ? (json['amountCents'] as num).toInt()
          : null,
      currency: json['currency'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'type': type == PricingType.free
          ? 'free'
          : type == PricingType.total
              ? 'total'
              : 'per_person',
      if (amountCents != null) 'amountCents': amountCents,
      if (currency != null) 'currency': currency,
    };
  }

  bool get isFree => type == PricingType.free || (amountCents ?? 0) == 0;

  String get displayPrice {
    if (isFree) return 'Free';
    final dollars = (amountCents ?? 0) / 100;
    return '\$${dollars.toStringAsFixed(0)}';
  }
}

/// GameSummary contains essential game details for list views
class GameSummary {
  final String id;
  final GameCategory category;
  final String? title;
  final String? description;
  final Location location;
  final DateTime startTime;
  final int durationMinutes;
  final int maxParticipants;
  final int signupCount;
  final Pricing pricing;
  final DateTime signupDeadline;
  final SkillLevel skillLevel;
  final GameStatus status;
  final String? userParticipationStatus;

  GameSummary({
    required this.id,
    required this.category,
    this.title,
    this.description,
    required this.location,
    required this.startTime,
    required this.durationMinutes,
    required this.maxParticipants,
    required this.signupCount,
    required this.pricing,
    required this.signupDeadline,
    required this.skillLevel,
    required this.status,
    this.userParticipationStatus,
  });

  factory GameSummary.fromJson(Map<String, dynamic> json) {
    return GameSummary(
      id: json['id'] as String,
      category: GameCategory.fromString(json['category'] as String),
      title: json['title'] as String?,
      description: json['description'] as String?,
      location: Location.fromJson(json['location'] as Map<String, dynamic>),
      startTime: DateTime.parse(json['startTime'] as String),
      durationMinutes: json['durationMinutes'] != null
          ? (json['durationMinutes'] as num).toInt()
          : 0,
      maxParticipants: json['maxParticipants'] != null
          ? (json['maxParticipants'] as num).toInt()
          : 0,
      signupCount: json['signupCount'] != null
          ? (json['signupCount'] as num).toInt()
          : 0,
      pricing: Pricing.fromJson(json['pricing'] as Map<String, dynamic>),
      signupDeadline: DateTime.parse(json['signupDeadline'] as String),
      skillLevel: SkillLevel.fromString(json['skillLevel'] as String),
      status: GameStatus.fromString(json['status'] as String),
      userParticipationStatus: json['userParticipationStatus'] as String?,
    );
  }

  bool get isFull => signupCount >= maxParticipants;
  bool get isUserJoined => userParticipationStatus != null &&
      (userParticipationStatus == 'confirmed' || userParticipationStatus == 'waitlist');
}

/// Game represents a full game with all details including participants
class Game {
  final String id;
  final User? owner;
  final GameCategory category;
  final String? title;
  final String? description;
  final Location location;
  final DateTime startTime;
  final int durationMinutes;
  final int maxParticipants;
  final List<dynamic> confirmedParticipants; // Confirmed participants (up to max)
  final List<dynamic> waitlist; // Waitlisted participants (beyond max)
  final Pricing pricing;
  final DateTime signupDeadline;
  final DateTime? dropDeadline;
  final SkillLevel skillLevel;
  final String? notes;
  final GameStatus status;
  final DateTime? cancelledAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  Game({
    required this.id,
    this.owner,
    required this.category,
    this.title,
    this.description,
    required this.location,
    required this.startTime,
    required this.durationMinutes,
    required this.maxParticipants,
    required this.confirmedParticipants,
    required this.waitlist,
    required this.pricing,
    required this.signupDeadline,
    this.dropDeadline,
    required this.skillLevel,
    this.notes,
    required this.status,
    this.cancelledAt,
    required this.createdAt,
    required this.updatedAt,
  });

  // Computed properties
  int get currentParticipants => confirmedParticipants.length;
  int get waitlistCount => waitlist.length;

  factory Game.fromJson(Map<String, dynamic> json) {
    return Game(
      id: json['id'] as String,
      owner: json['owner'] != null
          ? User.fromJson(json['owner'] as Map<String, dynamic>)
          : null,
      category: GameCategory.fromString(json['category'] as String),
      title: json['title'] as String?,
      description: json['description'] as String?,
      location: Location.fromJson(json['location'] as Map<String, dynamic>),
      startTime: DateTime.parse(json['startTime'] as String),
      durationMinutes: json['durationMinutes'] != null
          ? (json['durationMinutes'] as num).toInt()
          : 0,
      maxParticipants: json['maxParticipants'] != null
          ? (json['maxParticipants'] as num).toInt()
          : 0,
      confirmedParticipants: json['confirmedParticipants'] != null
          ? (json['confirmedParticipants'] as List<dynamic>)
          : [],
      waitlist: json['waitlist'] != null
          ? (json['waitlist'] as List<dynamic>)
          : [],
      pricing: Pricing.fromJson(json['pricing'] as Map<String, dynamic>),
      signupDeadline: DateTime.parse(json['signupDeadline'] as String),
      dropDeadline: json['dropDeadline'] != null
          ? DateTime.parse(json['dropDeadline'] as String)
          : null,
      skillLevel: SkillLevel.fromString(json['skillLevel'] as String),
      notes: json['notes'] as String?,
      status: GameStatus.fromString(json['status'] as String),
      cancelledAt: json['cancelledAt'] != null
          ? DateTime.parse(json['cancelledAt'] as String)
          : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      if (owner != null) 'owner': owner!.toJson(),
      'category': category.toApiString(),
      if (title != null) 'title': title,
      if (description != null) 'description': description,
      'location': location.toJson(),
      'startTime': startTime.toIso8601String(),
      'durationMinutes': durationMinutes,
      'maxParticipants': maxParticipants,
      'confirmedParticipants': confirmedParticipants,
      'waitlist': waitlist,
      'pricing': pricing.toJson(),
      'signupDeadline': signupDeadline.toIso8601String(),
      if (dropDeadline != null) 'dropDeadline': dropDeadline!.toIso8601String(),
      'skillLevel': skillLevel.toString().split('.').last,
      if (notes != null) 'notes': notes,
      'status': status.toString().split('.').last,
      if (cancelledAt != null) 'cancelledAt': cancelledAt!.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  bool get isFull => currentParticipants >= maxParticipants;
  bool get hasWaitlist => waitlistCount > 0;
}

/// CreateGameRequest for creating a new game
class CreateGameRequest {
  final GameCategory category;
  final String? title;
  final String? description;
  final Location location;
  final DateTime startTime;
  final int durationMinutes;
  final int maxParticipants;
  final Pricing pricing;
  final DateTime? signupDeadline;
  final SkillLevel skillLevel;
  final String? notes;

  CreateGameRequest({
    required this.category,
    this.title,
    this.description,
    required this.location,
    required this.startTime,
    required this.durationMinutes,
    required this.maxParticipants,
    required this.pricing,
    this.signupDeadline,
    this.skillLevel = SkillLevel.all,
    this.notes,
  });

  Map<String, dynamic> toJson() {
    String formatDateTime(DateTime dt) {
      final offset = dt.timeZoneOffset;
      final offsetSign = offset.isNegative ? '-' : '+';
      final offsetHours = offset.inHours.abs().toString().padLeft(2, '0');
      final offsetMinutes = (offset.inMinutes.abs() % 60).toString().padLeft(2, '0');

      return '${dt.year.toString().padLeft(4, '0')}-${dt.month.toString().padLeft(2, '0')}-${dt.day.toString().padLeft(2, '0')}'
             'T${dt.hour.toString().padLeft(2, '0')}:${dt.minute.toString().padLeft(2, '0')}:${dt.second.toString().padLeft(2, '0')}'
             '$offsetSign$offsetHours:$offsetMinutes';
    }

    return {
      'category': category.toApiString(),
      if (title != null && title!.isNotEmpty) 'title': title,
      if (description != null && description!.isNotEmpty) 'description': description,
      'location': location.toJson(),
      'startTime': formatDateTime(startTime),
      'durationMinutes': durationMinutes,
      'maxParticipants': maxParticipants,
      'pricing': pricing.toJson(),
      if (signupDeadline != null) 'signupDeadline': formatDateTime(signupDeadline!),
      'skillLevel': skillLevel.toString().split('.').last,
      if (notes != null && notes!.isNotEmpty) 'notes': notes,
    };
  }
}
